test_othermlpackages/package_test_common.py

```diff
@@ -29,42 +33,88 @@ def fetch_adult():
     X_scaled = sc.fit_transform(X)
     X_scaled = pd.DataFrame(X_scaled, columns=X.columns)
 
-    return X_scaled, Y, A
+    X_train, X_test, Y_train, Y_test, A_train, A_test = \
+        train_test_split(X_scaled, Y, A,
+                         test_size=0.3,
+                         random_state=12345,
+                         stratify=Y)
+
+    # Ensure indices are aligned
+    X_train = X_train.reset_index(drop=True)
+    A_train = A_train.reset_index(drop=True)
+    X_test = X_test.reset_index(drop=True)
+    A_test = A_test.reset_index(drop=True)
+
+    return X_train, Y_train, A_train, X_test, Y_test, A_test
 
 
 def run_expgrad_classification(estimator, moment):
     """Run classification test with ExponentiatedGradient."""
-    X, Y, A = fetch_adult()
+    X_train, Y_train, A_train, X_test, Y_test, A_test = fetch_adult()
+    verification_moment = copy.deepcopy(moment)
+
+    unmitigated = copy.deepcopy(estimator)
+    unmitigated.fit(X_train, Y_train)
 
     expgrad = ExponentiatedGradient(
         estimator,
         constraints=moment)
-    expgrad.fit(X, Y, sensitive_features=A)
+    expgrad.fit(X_train, Y_train, sensitive_features=A_train)
 
     assert expgrad.n_oracle_calls_ > 1
     assert len(expgrad.predictors_) > 1
 
+    verification_moment.load_data(X_test, Y_test, sensitive_features=A_test)
+    gamma_unmitigated = verification_moment.gamma(lambda x: unmitigated.predict(x))
+    gamma_mitigated = verification_moment.gamma(lambda x: expgrad.predict(x))
+
+    for idx in gamma_mitigated.index:
+        assert abs(gamma_mitigated[idx]) <= abs(gamma_unmitigated[idx]), "Checking {0}".format(idx)
+
 
 def run_gridsearch_classification(estimator, moment):
     """Run classification test with GridSearch."""
-    X, Y, A = fetch_adult()
+    X_train, Y_train, A_train, X_test, Y_test, A_test = fetch_adult()
+    verification_moment = copy.deepcopy(moment)
 
-    num_predictors = 5
+    unmitigated = copy.deepcopy(estimator)
+    unmitigated.fit(X_train, Y_train)
+
+    num_predictors = 11
     gs = GridSearch(
         estimator,
         constraints=moment,
         grid_size=num_predictors)
-    gs.fit(X, Y, sensitive_features=A)
+    gs.fit(X_train, Y_train, sensitive_features=A_train)
 
     assert len(gs.predictors_) == num_predictors
 
+    verification_moment.load_data(X_test, Y_test, sensitive_features=A_test)
+    gamma_unmitigated = verification_moment.gamma(lambda x: unmitigated.predict(x))
+    gamma_mitigated = verification_moment.gamma(lambda x: gs.predict(x))
+
+    for idx in gamma_mitigated.index:
+        assert abs(gamma_mitigated[idx]) <= abs(gamma_unmitigated[idx]), "Checking {0}".format(idx)
```

Comment:
@MiroDudik is this a sensible way to check that the reductions have actually done what they're supposed to?

Reply:
You could run expgrad until convergence and check if the violation is below the bound (assuming that it actually converges as opposed to hitting the iteration limit)

For grid search there are no guarantees like that.